vllm/v1/engine/core_client.py

```diff
@@ -383,72 +362,161 @@ def sigusr1_handler(signum, frame):
+        local_start_index = parallel_config.data_parallel_rank_local
+
+        # SPMD mode is where there is an LLM instance per DP rank and one
+        # core engine per LLM, see examples/offline_inference/data_parallel.py.
+        spmd_mode = local_start_index is not None
+        if spmd_mode:
+            assert local_engine_count == 1
+            self.core_engines = [
+                CoreEngine(index=local_start_index, local=True)
+            ]
+        else:
+            assert start_index == 0
+            local_start_index = 0
+            self.core_engines = [
+                CoreEngine(index=i, local=(i < local_engine_count))
+                for i in range(parallel_config.data_parallel_size)
+            ]
+
+        input_address, output_address = self._get_zmq_addresses(
+            parallel_config, spmd_mode)
+
+        # Create input and output sockets.
+        self.input_socket = self.resources.input_socket = make_zmq_socket(
+            self.ctx, input_address, zmq.ROUTER, bind=True)
+
+        self.resources.output_socket = make_zmq_socket(self.ctx,
+                                                       output_address,
+                                                       zmq.constants.PULL)
+        # Start local engines.
+        if local_engine_count:
+            # In server mode, start_index and local_start_index will both be 0.
+            self.resources.local_engine_manager = CoreEngineProcManager(
+                EngineCoreProc.run_engine_core,
+                vllm_config=vllm_config,
+                executor_class=executor_class,
+                log_stats=log_stats,
+                input_address=input_address,
+                on_head_node=True,
+                local_engine_count=local_engine_count,
+                start_index=start_index,
+                local_start_index=local_start_index)
+
+        self.core_engine = self.core_engines[0]
 
         # Wait for engine core process(es) to start.
-        self._wait_for_engine_startup()
+        self._wait_for_engine_startup(output_address, parallel_config)
 
         self.utility_results: dict[int, AnyFuture] = {}
 
-    def _wait_for_engine_startup(self):
+    @staticmethod
+    def _get_zmq_addresses(parallel_config: ParallelConfig,
+                           spmd_mode: bool) -> tuple[str, str]:
+        """Returns (input_address, output_address)."""
+        dp_size = parallel_config.data_parallel_size
+        local_engine_count = parallel_config.data_parallel_size_local
+
+        if local_engine_count == dp_size or spmd_mode:
+            input_address = get_open_zmq_ipc_path()
+            output_address = get_open_zmq_ipc_path()
+        else:
+            host = parallel_config.data_parallel_master_ip
+            input_port = parallel_config.data_parallel_rpc_port
+            output_port = get_open_port()
+            input_address = f"tcp://{host}:{input_port}"
+            output_address = f"tcp://{host}:{output_port}"
```

Comment:
This requires a bit more care for IPv6. If the address is IPv6, you need square brackets around `{host}`.

Similar code here: https://github.com/vllm-project/vllm/blob/34b2cf3b3366443e6b68a43572aee1800f6f2b90/vllm/distributed/device_communicators/shm_broadcast.py#L238-L241

Note that for zmq, the socket also needs an extra flag set to support IPv6: 

https://github.com/vllm-project/vllm/blob/34b2cf3b3366443e6b68a43572aee1800f6f2b90/vllm/distributed/device_communicators/shm_broadcast.py#L303-L304

though the latter is probably a fix to `make_zmq_socket`, which isn't new code in this PR. I can fix that part separately. At least for this PR, ensuring the zmq URLs are correct is enough.

Reply:
Now addressed, thanks!